Forced translation

pedrodiascoelho opened the pull request:
This PR serves the purpose of delivering the code related to the second Milestone of the marian-cef project. I opened the PR to easy the discussion and so that you can leave comments if you have any. @snukky @kpu 

Here you can find all the needed scripts to execute an end-to-end training, of a model capable of handling forced translation.

We provide a description of how to execute the end-to-end pipeline [here](https://github.com/marian-cef/marian-examples/tree/forced-translation/forced-translation#readme), we detailed the method used and give a walkthrough regarding that same pipeline [here](https://github.com/marian-cef/marian-examples/blob/forced-translation/forced-translation/docs/Pipeline.md), present the experiments that we conducted [here](https://github.com/marian-cef/marian-examples/blob/forced-translation/forced-translation/docs/Experiments.md), and also show how to execute individually the core scripts so that you can use this technique into your own training pipelines [here](https://github.com/marian-cef/marian-examples/blob/forced-translation/forced-translation/scripts/README.md).

For now I’ve linked the marian-cef repo as the repository to use because the end-2-end pipeline is prepared to used concatenation as a way to combine factors and lemmas, as that is what we did in our experiments. I am still overviewing @snukky  comments on the marian-dev PR, and as soon as all the issues are solved, I’ll make the code linked with the regular marian by default. Nevertheless the current marian-dev can be used to run this pipeline, just remove the `—factor-dim-emb` and `--factors-combine` options from the training task,and it will use sum as the way to combine lemmas and factors.

Finally we had some issues with the human eval for `en-ro` and `en-no`, but we hope to deliver those results as soon as possible.

Any questions feel free to reach me.

pedrocoelhounbabel committed:
add forced-translation examples

pedrocoelhounbabel committed:
Update .gitignore

pedrocoelhounbabel committed:
Fix some factor nos being correctly applied

tomsbergmanis committed:
Update README.md

tomsbergmanis committed:
Update Experiments.md

pedrocoelhounbabel committed:
remove glossary tokenization

pedrocoelhounbabel committed:
add eval scripts

kpu commented:
As a very preliminary thing, all files should end with a newline.  (And there's no need for a lot of them in forced-translation/test_data/download_data.sh)

pedrocoelhounbabel committed:
Fix end of files

qianqianzhu commented on forced-translation/README.md:
```diff
@@ -0,0 +1,69 @@
+
+## Install
+
+Download and compile [Marian-NMT](https://github.com/marian-cef/marian-dev). For now it is needed to install the linked marian-cef version of Marian to use all the capabilities related to input factors, but as soon as the code is merged to marian-dev the link will be updated. Once Marian is installed add the path to the marian executables folder in the `.env` file.
```
One quick question: what are the compilation options for Marian here? I am trying to run your script.

pedrodiascoelho replied:
I compiled it with the following options: 
```
cmake -DUSE_SENTENCEPIECE=ON -DUSE_FBGEMM=on     -DCOMPILE_CPU=on -DCOMPILE_TESTS=ON -DCOMPILE_EXAMPLES=ON -DCUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-10.1 ..
```
just because that is how I am used to compile marian. This repo should not require any extra compilation option rather than CUDA related stuff

qianqianzhu replied:
Many thanks!

pedrocoelhounbabel committed:
Update EXPERIMENTS.md

    Added information regarding the two new testsets used for en-ro and en-nb, that contained a specific domained, annotated with a glossary with terms only specific to that domain to mimic better the Tilde's ATS testset and glossary, used for en-lv and en-de. Also added the human evaluation results for this two LPs.

qianqianzhu commented on forced-translation/README.md:
````diff
@@ -0,0 +1,69 @@
+### Trigger pipeline
+After everything is ready to go just execute:
+```
+./run-me.sh
````
The `run-me.sh` script ran for more than 12 hours and it is still running (with two GeForce RTX 2080 Ti). How long does it normally take? (If the script takes very long, maybe it is better to add some note/indicator here)

pedrodiascoelho replied:
Thank you for the feedback!
Updated the readme with some time estimations. Let me know if that goes in accordance to what you have experienced!

pedrocoelhounbabel committed:
Add time estimations to run end-2-end pipeline

tomsbergmanis committed:
Update Experiments.md
